File: exp/pproc/postHocTest.m
function [pv, summary] = postHocTest(data, varargin)
%MULTCOMP Pairwise comparison of algorithms on a set of benchmarking
%   functions. Performs Friedman posthoc test and adjusts pairwise p-values
%   error by Bergmann-Hommel correction.
%   Currently implemented as a wrapper over an R script.
%
%   Input:
%     data - a table with benchmarking function as rows and algorithms as
%            variables containing best achieved fitness values
%     posthoc_test - 'friedman' (default) | 'quade' | 'aligned ranks'
%     (Friedman Aligned Ranks)
%     correction - 'bergmann' (default) | 'shaffer' | 'bonferroni'
%
%   Output:
%     pv      - a table of adjusted p-values with algorithms in columns and
%            rows
%     summary - test summary such as Friedman's mean ranks
%
%   TODO:
%     [ ] comparison of multiple algorithms with a control

  Rpath = fullfile('exp', 'pproc', 'R');
  Rscript = 'postHocTest.R';
  fin = [tempname, '.in.csv'];
  fout_pv = [tempname, '.pv.out.csv'];
  fout_sum = [tempname, '.sum.out.csv'];

  if nargin >= 2
    if ismember(varargin{1}, ...
      {'friedman', 'wilcoxon', 'quade', 'aligned ranks'})
      test = varargin{1};
    else
      error('Unrecognized test: ''%s''', varargin{1});
    end
  else
    test = 'friedman';
  end

  if nargin >= 3
    if ismember(varargin{2}, ...
      {'bergmann', 'shaffer', 'bonferroni'})
      corr = varargin{2};
    else
      error('Unrecognized correction type: ''%s''', varargin{2});
    end
  else
    corr = 'bergmann';
  end

  % write the input file with floats in scientific notation
  dlmwrite(fin, data, 'precision', '%e', 'delimiter', ',');

  args = sprintf(' -i "%s" -p "%s" -s "%s" -t "%s" -c "%s"', ...
    fin, fout_pv, fout_sum, test, corr);
  cmd = [fullfile(Rpath, Rscript), args];

  % execute the script and check exit code
  [status, cmdout] = system(cmd);

  if status ~= 0
    delete(fin);
    if exist(fout_pv, 'file'), delete(fout_pv), end
    if exist(fout_sum, 'file'), delete(fout_sum), end

    error('Script %s ended with error status %d and output:\n%s', ...
      Rscript, status, cmdout);
  end
  
  % read the script's output or outputs
  pv = dlmread(fout_pv, ',');
  summary = dlmread(fout_sum, ',');

  % clean up
  delete(fin);
  if exist(fout_pv, 'file'), delete(fout_pv), end
  if exist(fout_sum, 'file'), delete(fout_sum), end
end

File: exp/pproc/R/postHocTest.R
#!/usr/bin/env Rscript
#
# A pairwise multi-comparison of algorithms with procedures described in
# Demsar (2006) and Garcia & Herrera (2008).
#
# Dependencies:
# * scmamp
# * Rgraphviz (for correct installation of the former)
# * optparse
#
# How to install dependencies in an R session:
# > install.packages("optparse")
# > # moved from CRAN to bioconductor
# > source("http://www.bioconductor.org/biocLite.R")
# > biocLite("Rgraphviz")
# > install.packages("scmamp")

suppressPackageStartupMessages(library("optparse"))
suppressPackageStartupMessages(library("scmamp"))

# define command line arguments
option_list <- list(
                    make_option(c("-i", "--input"), default="algs.csv",
                                action="store",
                                help="Input file (default \"%default\")"),
                    make_option(c("-p", "--padj_out"), action="store",
                                default="",
                                help=paste0("Output file for adjusted p-vals ",
                                            "(default stdout).")),
                    make_option(c("-s", "--summary_out"), action="store",
                                default="",
                                help=paste0("Output file for the test summary ",
                                            "(default stdout).")),
                    make_option(c("-t", "--posthoc_test"), action="store",
                                default="friedman",
                                help=paste0("Posthoc test. One of ",
                                            "\"friedman\" (default), ",
                                            "\"wilcoxon\", ",
                                            "\"quade\", ",
                                            "\"aligned ranks\".")),
                    make_option(c("-c", "--correction"), action="store",
                                default="bergmann",
                                help=paste0("P-values correction method. One of ",
                                            "\"bergmann\" (default), ",
                                            "\"shaffer\", or \"bonferrroni\"."))

               )

opt_parser = OptionParser(usage="usage: %prog [options]",
                          option_list=option_list, 
                          add_help_option=TRUE, prog=NULL, description="",
                          epilogue="")

# parse command line arguments
opts <- parse_args(opt_parser, print_help_and_exit=TRUE)

stopifnot(any(opts$posthoc_test == c("friedman", "quade",
                                     "wilcoxon", "aligned ranks")))
stopifnot(any(opts$correction == c("bergmann", "shaffer", "bonferroni")))

# read data
fvalues <- read.table(opts$input, header=FALSE, sep=",")
fvalues <- (max(fvalues)-fvalues)/(max(fvalues)-min(fvalues))

# perform the posthoc test
res <- postHocTest(data=fvalues, test=opts$posthoc_test,
                   control=NULL, use.rank=TRUE,
                   correct=opts$correction)

# output adjusted p-values
write.table(res$corrected.pval, opts$padj_out, na="NaN", sep=",",
            quote=FALSE, eol="\n", row.names=FALSE, col.names=FALSE)

# output summary
write.table(res$summary, opts$summary_out, na="NaN", sep=",",
            quote=FALSE, eol="\n", row.names=FALSE, col.names=FALSE)
